File: app/src/main/java/com/blog/demo/custom/widget/PhoneTextView.java
package com.blog.demo.custom.widget;

import android.content.Context;
import android.text.Editable;
import android.text.InputFilter;
import android.text.InputType;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.widget.EditText;

import com.blog.demo.LogTool;

public class PhoneTextView extends EditText {

    public PhoneTextView(Context context) {
        this(context, null);
    }

    public PhoneTextView(Context context, AttributeSet attrs) {
        super(context, attrs);

        setFilters(new InputFilter[] {new InputFilter.LengthFilter(13)});
        setInputType(InputType.TYPE_CLASS_NUMBER);

        addTextChangedListener(new PhoneTextWatcher(this));
    }

    static class PhoneTextWatcher implements TextWatcher {
        private static final char SPACE_CHAR = ' ';

        private EditText mEditText;

        PhoneTextWatcher(EditText editText) {
            this.mEditText = editText;
        }

        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            int selection = start + count;

            char[] array = new char[s.length()];
            int len = 0;
            int space = 0;
            for (int index = 0; index < s.length(); index++) {
                if (s.charAt(index) == SPACE_CHAR) {
                    if (index < selection) {
                        ++space;
                    }
                } else {
                    array[len++] = s.charAt(index);
                }
            }
            setPhoneText(new String(array, 0, len), selection - space);
        }

        private void setPhoneText(String phoneText, int selection) {
            LogTool.logi("PhoneTextView", phoneText + ", selection = " + selection);
            String text = "";
            int space = 0;
            for (int index = 0; index < phoneText.length(); index++) {
                if (isEmptyPlace(text.length())) {
                    text += SPACE_CHAR;
                    if (index < selection) {
                        ++space;
                    }
                }
                text += phoneText.charAt(index);
            }

            mEditText.removeTextChangedListener(this);
            mEditText.setText(text);
            mEditText.setSelection(selection + space);
            mEditText.addTextChangedListener(this);
        }

        private boolean isEmptyPlace(int index) {
            return 3 == index || 8 == index;
        }

        @Override
        public void afterTextChanged(Editable s) {
        }
    }

}

File: app/src/main/java/com/blog/demo/custom/widget/PhoneTextView.kt
package com.blog.demo.custom.widget

import android.content.Context
import android.text.Editable
import android.text.InputFilter
import android.text.InputFilter.LengthFilter
import android.text.InputType
import android.text.TextWatcher
import android.util.AttributeSet
import android.widget.EditText
import com.blog.demo.LogTool.logi

class PhoneTextView(context: Context, attrs: AttributeSet?, defStyleAttr: Int) :
        androidx.appcompat.widget.AppCompatEditText(context, attrs, defStyleAttr) {

    constructor(context: Context) : this(context, null)

    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)

    init {
        filters = arrayOf<InputFilter>(LengthFilter(13))
        inputType = InputType.TYPE_CLASS_NUMBER

        addTextChangedListener(PhoneTextWatcher(this))
    }

    internal class PhoneTextWatcher(private val mEditText: EditText) : TextWatcher {

        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
        }

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            val selection = start + count
            val array = CharArray(s.length)
            var len = 0
            var space = 0
            for (index in s.indices) {
                if (s[index] == SPACE_CHAR) {
                    if (index < selection) {
                        ++space
                    }
                } else {
                    array[len++] = s[index]
                }
            }
            setPhoneText(String(array, 0, len), selection - space)
        }

        private fun setPhoneText(phoneText: String, selection: Int) {
            logi("PhoneTextView", "$phoneText, selection = $selection")
            var text = ""
            var space = 0
            for (index in phoneText.indices) {
                if (isEmptyPlace(text.length)) {
                    text += SPACE_CHAR
                    if (index < selection) {
                        ++space
                    }
                }
                text += phoneText[index]
            }
            mEditText.removeTextChangedListener(this)
            mEditText.setText(text)
            mEditText.setSelection(selection + space)
            mEditText.addTextChangedListener(this)
        }

        private fun isEmptyPlace(index: Int): Boolean {
            return 3 == index || 8 == index
        }

        override fun afterTextChanged(s: Editable) {}

        companion object {
            private const val SPACE_CHAR = ' '
        }
    }

}